bluepymm/templates/cell_template_neurodamus_sbo.jinja2

```diff
@@ -0,0 +1,274 @@
+  objref this, CellRef, segCounts, ASCIIrpt, synHelperList, synlist
+
+  public all, somatic, apical, axonal, basal, myelinated, APC
+  objref all, somatic, apical, axonal, basal, myelinated, APC
+
+
+obfunc getCell(){
+        return this
+}
+
+obfunc getCCell(){
+	return CellRef
+}
+proc setCCell(){
+       CellRef = $o1
+}
+
+//-----------------------------------------------------------------------------------------------
+
+/*!
+ * When clearing the model, the circular reference between Cells and CCells must be broken so the
+ * entity watching reference counts can work.
+ */
+proc clear() { localobj nil
+    CellRef = nil
+}
+
+
+
+/*!
+ * @param $o1 NetCon source (can be nil)
+ * @param $o2 Variable where generated NetCon will be placed
+ */
+proc connect2target() { //$o1 target point process, $o2 returned NetCon
+  soma $o2 = new NetCon(&v(1), $o1)
+  $o2.threshold = -30
+}
+
+
+proc init(/* args: morphology_dir, morphology_name */) {
+  all = new SectionList()
+  apical = new SectionList()
+  axonal = new SectionList()
+  basal = new SectionList()
+  somatic = new SectionList()
+  myelinated = new SectionList()
+
+  synHelperList = new List()                                                     
+  synlist = new List()     
+
+  //For compatibility with BBP CCells
+  CellRef = this
+
+  forall delete_section()
+
+  gid = $1
+
+  if(numarg() >= 3) {
+    load_morphology($s2, $s3)
+  } else {
+  {%- if morphology %}
+    load_morphology($s2, "{{morphology}}")
+  {%- else %}
+    execerror("Template {{template_name}} requires morphology name to instantiate")
+  {%- endif %}
+  }
+
+  geom_nseg()
+  indexSections()
+  {%- if replace_axon %}
+    replace_axon()
+  {%- endif %}
+  insertChannel()
+  biophys()
+
+  // Initialize channel_seed_set to avoid accidents
+  channel_seed_set = 0
+  // Initialize random number generators
+  re_init_rng()
+}
+
+/*!
+ * Assign section indices to the section voltage value.  This will be useful later for serializing
+ * the sections into an array.  Note, that once the simulation begins, the voltage values will revert to actual data again.
+ *
+ * @param $o1 Import3d_GUI object
+ */
+proc indexSections() { local index
+    index = 0
+    forsec all {
+        v(0.0001) = index
+        index = index +1
+    }
+}
+
+func getThreshold() { return 0.0 }
+
+proc load_morphology(/* morphology_dir, morphology_name */) {localobj morph, import, sf, extension, commands, pyobj
+  strdef morph_path
+  sprint(morph_path, "%s/%s", $s1, $s2)  sf = new StringFunctions()
+  extension = new String()  sscanf(morph_path, "%s", extension.s)
+
+  // TODO fix the `-3` here.
+  sf.right(extension.s, sf.len(extension.s)-3)
+  
+  if( strcmp(extension.s, ".asc") == 0 ) {
+     morph = new Import3d_Neurolucida3()
+  } else if( strcmp(extension.s, ".swc" ) == 0) {
+     morph = new Import3d_SWC_read()
+  } else if( strcmp(extension.s, ".h5") == 0 ) {
+    if(nrnpython ("from morphio_wrapper import MorphIOWrapper") == 1) {
```

Comment:
One question: don't we need to have this dependency in the setup.py?

Reply:
It is up to the user to have the correct dependencies to use the file that would be outputted with this piece of code. We don't have to add a dependency we don't need to use BPMM

Reply:
This template is just used to create a new file, BPMM will not try to execute the code in it, afaik

Reply:
I see. Although the code is called in the prepare step, the next step (run) will require the execution and it will give a runtime error on h5 morphologies.

That file that gets imported in the hoc is not a package on pip, it's one of the inner modules of Neurodamus and Neurodamus is not on pip. https://github.com/BlueBrain/neurodamus/blob/main/neurodamus/morphio_wrapper.py

I'm afraid this is going to be an issue for the packages that need to run the h5 loading hoc template e.g. bluepymm, bluecellulab, bluepyopt.

Reply:
I don't know if there is plan to put neurodamus on pypi. Let me ask on the HPCTM-1754 ticket.

Reply:
no, aurelien is correct, bluepymm just generate hoc files to be consumed by neurodamus, it doe snot run them. In any case, I'm in charge of MM, and I'm not using this code ;P

Reply:
I'm a bit confused. Isn't the `run` step of bluepymm used anymore? Is the `select` step also not used?

Reply:
it is, it does not run from hoc files, but via 'bluepyemodel' type of code using bluepyopt, etc...

Reply:
Yes, good we're on the same page. So whichever software (bluecellulab, bluepyopt) that needs to run the hoc with an h5 morphology will require the inner `morphio_wrapper`  module of neurodamus, right?

Reply:
it looks like it, but I don't know more than you on that ;P

Reply:
It seems there is a plan to enable h5 morph support in NEURON. Once that happens bluecellulab and bluepyopt can also load those cells.